### PR TITLE
Add a utility for finding minimal topological sorts

### PR DESCRIPTION
**Pull request by tlively**
Reuse the code implementing Kahn's topological sort algorithm with a new
configuration that uses a min-heap to always choose the best available
element.

Also add wrapper utilities that can find topological sorts of graphs
with arbitrary element types, not just indices.

### CODE REVIEW

File: src/support/topological_orders.h [hunk 1]
```diff
@@ -36,7 +38,8 @@ struct TopologicalOrders {
 
   // Takes an adjacency list, where the list for each vertex is a sorted list of
   // the indices of its children, which will appear after it in the order.
-  TopologicalOrders(const std::vector<std::vector<size_t>>& graph);
+  TopologicalOrders(const std::vector<std::vector<size_t>>& graph)
+    : TopologicalOrders(graph, false) {}
```
Comment (kripken) [inline src/support/topological_orders.h:42]: Maybe an enum for this?

File: src/support/topological_orders.h [hunk 2]
```diff
@@ -78,19 +88,86 @@ struct TopologicalOrders {
...
 
+// A utility for finding a single topological order of a graph.
+struct TopologicalSort : private TopologicalOrders {
+  TopologicalSort(const std::vector<std::vector<size_t>>& graph)
+    : TopologicalOrders(graph, false) {}
+
+  const std::vector<size_t>& operator*() const {
+    return TopologicalOrders::operator*();
+  }
+};
```
Comment (kripken) [inline src/support/topological_orders.h:115]: This constructor is identical to the parent, and this operator doesn't seem to modify anything either - I am sure I'm missing something?

File: test/gtest/topological-orders.cpp [hunk 3]
```diff
@@ -99,3 +99,55 @@ TEST(TopologicalOrdersTest, Diamond) {
...
+  graph[1].push_back(0);
+  std::vector<size_t> expected{1, 0, 2};
+  EXPECT_EQ(*MinTopologicalSort(graph), expected);
+}
+
+TEST(MinTopologicalSortTest, TwoBeforeOne) {
+  Graph graph(3);
+  graph[2].push_back(1);
+  std::vector<size_t> expected{0, 2, 1};
+  EXPECT_EQ(*MinTopologicalSort(graph), expected);
```
Comment (kripken) [inline test/gtest/topological-orders.cpp:134]: Maybe add comments to these, e.g. for this one IIUC the point is to show that 2 is before 1, and the unconstrained 0 is smaller so it appears before them both.

**Review by kripken — APPROVED** [review]

File: src/support/topological_orders.h (hunk 2 above)
Comment (tlively) [reply src/support/topological_orders.h:115]: Nope, this is providing a small subset of the functionality of the superclass, packaged under a different name that is more appropriate for the common case of needing just one topological sort.

Comment (kripken) [reply src/support/topological_orders.h:115]: Ah, thanks... I was missing the word "private" on line 108. Makes sense now.

**Pull request merged**